feat(no-shadow-native-events): initial implementation (#2558)
vuejs/eslint-plugin-vue: docs/rules/no-shadow-native-events.md

---
pageClass: rule-details
sidebarDepth: 0
title: vue/no-shadow-native-events

> description: disallow `emits` which would shadow native html events

Reviewer: Can you please run `npm run update` to update the rule docs with the description from the rule source itself?

Author: done

---
# vue/no-shadow-native-events

> disallow `emits` which would shadow native HTML events

- :exclamation: <badge text="This rule has not been released yet." vertical="middle" type="error"> _**This rule has not been released yet.**_ </badge>

## :book: Rule Details

> This rule reports emits that shadow native HTML events. (The `emits` option is a new in Vue.js 3.0.0+)

Reviewer: This is not relevant for this rule, since `$emit()` calls are also reported.

```suggestion
This rule reports emits that shadow native HTML events.
```

Using native event names for emits can lead to incorrect assumptions about an emit and cause confusion. This is caused by Vue emits behaving differently from native events. E.g. :

- The payload of an emit can be chosen arbitrarily
- Vue emits do not bubble, while most native events do
- [Event modifiers](https://vuejs.org/guide/essentials/event-handling.html#event-modifiers) only work on HTML events or when the original event is re-emitted as emit payload.

> - When the native event is remitted the `event.target` might not match the actual event-listeners location.

Reviewer: ```suggestion
- When the native event is re-emitted, the `event.target` might not match the actual event-listeners location.
```

---
pageClass: rule-details
sidebarDepth: 0
title: vue/no-shadow-native-events
description: disallow the use of event names that collide with native web event names
---

# vue/no-shadow-native-events

> disallow the use of event names that collide with native web event names

- :exclamation: <badge text="This rule has not been released yet." vertical="middle" type="error"> _**This rule has not been released yet.**_ </badge>

## :book: Rule Details

This rule reports emits that shadow native HTML events.

> Using native event names for emits can lead to incorrect assumptions about an emit and cause confusion. This is caused by Vue emits behaving differently from native events. E.g. :

Reviewer: ```suggestion
Using native event names for emits can lead to incorrect assumptions about an emit and cause confusion. This is caused by Vue emits behaving differently from native events, e.g.:
```

- The payload of an emit can be chosen arbitrarily
- Vue emits do not bubble, while most native events do
- [Event modifiers](https://vuejs.org/guide/essentials/event-handling.html#event-modifiers) only work on HTML events or when the original event is re-emitted as emit payload.
- When the native event is re-emitted, the `event.target` might not match the actual event-listeners location.

> The rule is mostly aimed at developers of component libraries.

Reviewer: ```suggestion
Thus, you should avoid using native event names for Vue emits to avoid confusion, and consider emit names which communicate the users intent, if applicable.

The rule is mostly aimed at developers of component libraries.
```
